retnet/modeling_retnet.py

```diff
@@ -94,28 +141,35 @@ def __init__(self, config: RetNetConfig):
         self.register_buffer("angle", angle)
         self.register_buffer("decay", decay)
         self.recurrent_chunk_size = config.recurrent_chunk_size
-
+        self.cache = {
+            'parallel':{},
+            'recurrent':{},
+            'chunkwise':{}
+        }
     def forward(self,
                 slen,
                 forward_impl='parallel',
                 recurrent_chunk_size=None,
                 retention_mask=None,
                 get_decay_scale=True):
+        
         if forward_impl == 'recurrent':
+            if slen in self.cache[forward_impl]:
+                return self.cache[forward_impl][slen]
             sin = torch.sin(self.angle * (slen - 1))
             cos = torch.cos(self.angle * (slen - 1))
             retention_rel_pos = ((sin, cos), self.decay.view(1, -1, 1, 1).exp())
+            self.cache[forward_impl][slen] = retention_rel_pos
         elif forward_impl == 'chunkwise':
-            if recurrent_chunk_size is None:
-                recurrent_chunk_size = self.recurrent_chunk_size
+            
+            if recurrent_chunk_size is None:recurrent_chunk_size = self.recurrent_chunk_size
             index = torch.arange(slen).to(self.decay)
             sin = torch.sin(index[:, None] * self.angle[None, :])
             cos = torch.cos(index[:, None] * self.angle[None, :])
 
             block_index = torch.arange(recurrent_chunk_size).to(self.decay)
             mask = torch.tril(torch.ones(recurrent_chunk_size, recurrent_chunk_size)).to(self.decay)
-            mask = torch.masked_fill(block_index[:, None] - block_index[None, :], ~mask.bool(),
```

Comment:
I am using `yapf` formatter basically, with the hard-limit for column with 100 characters. Please see if you could not break the 100 character limit if it is not absolutely necessary to break it (I hardly doubt that is the case here)